libraries/node-core-library/src/FileError.ts

```diff
@@ -0,0 +1,159 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+import * as NodeJSPath from 'path';
+
+import { Path } from './Path';
+import { TypeUuid } from './TypeUuid';
+
+/**
+ * The format that the FileError message should conform to. The supported formats are:
+ *  - Unix: \<filePath\>:\<line\>:\<column\> - \<message\>
+ *  - VisualStudio: \<filePath\>(\<line\>,\<column\>) - \<message\>
+ *  - AzureDevOps: ##vso[task.logissue type=\<error|warning\>;sourcepath=\<filePath\>;linenumber=\<line\>;columnnumber=\<column\>;]\<message\>
+ *
+ * @public
+ */
+export type FileErrorFormat = 'Unix' | 'VisualStudio' | 'AzureDevOps';
+
+/**
+ * Provides options for the output message of a file error.
+ *
+ * @public
+ */
+export interface IFileErrorFormattingOptions {
+  /**
+   * The format for the error message. If no format is provided, format 'Unix' is used by default.
+   *
+   * @remarks If no format is specified and the environment variable TF_BUILD is set to 'True'
+   * (such as on Azure DevOps pipeline agents), the format 'AzureDevOps' is used by default.
+   */
+  format?: FileErrorFormat;
+
+  /**
+   * Whether or not the error is a warning. Defaults to false.
+   */
+  isWarning?: boolean;
+}
+
+const uuidFileError: string = '37a4c772-2dc8-4c66-89ae-262f8cc1f0c1';
+
+/**
+ * An `Error` subclass that should be thrown to report an unexpected state that specifically references
+ * a location in a file.
+ *
+ * @public
+ */
+export class FileError extends Error {
+  /**
+   * The path to the file that contains the error.
+   */
+  public readonly filePath: string;
+
+  /**
+   * The line number of the error in the target file.
+   */
+  public readonly line: number | undefined;
+
+  /**
+   * The column number of the error in the target file.
+   */
+  public readonly column: number | undefined;
+
+  /**
+   * Constructs a new instance of the {@link FileError} class.
+   *
+   * @param message - A message describing the error.
+   * @param filePath - The path to the file that contains the error.
+   * @param line - The line number of the error in the target file.
+   * @param column - The column number of the error in the target file.
+   */
+  public constructor(message: string, filePath: string, line?: number, column?: number) {
+    super(message);
+
+    this.filePath = Path.convertToSlashes(filePath);
+
+    // Error out if the file path is not an absolute path. Absolute paths provide targeted information
+    // about the error, and allow for navigation to the file from IDE consoles and CI systems.
+    if (!NodeJSPath.isAbsolute(this.filePath)) {
+      throw new Error('The filePath must be an absolute path.');
+    }
```

Comment:
I find the project-root-relative path more convenient than an absolute path

Reply:
Reporting errors using absolute paths allows you to click on them in VSCode and directly open the relevant file at the relevant line and column, which is _very useful_.